feat: add Turkish translation for useReducer (#428)
reactjs/tr.react.dev: src/content/reference/react/useReducer.md

---

> ## Başvuru Dokümanı {/*reference*/}

Reviewer: ```suggestion
## Referans {/*reference*/}
```

Author: Tesekkur ederim ancak linke bakarsan tavsiye edilen ceviriyi gorebilirsin.

https://github.com/reactjs/tr.react.dev/blob/main/TRANSLATION.md

Reference | Başvuru dokümanı
-- | --

  // ...
```

> [Aşağıda daha fazla örneğe bakın.](#usage)

Reviewer: ```suggestion
[Daha fazla örnek için aşağıya bakınız.](#usage)
```

> #### Parametereler {/*parameters*/}

Reviewer: ```suggestion
#### Parametreler {/*parameters*/}
```

> * `reducer`: State'in nasıl güncelleneceğini belirleyen reducer fonksiyonudur. Saf olmalıdır, state'i ve işlemi argüman olarak almalıdır ve bir sonraki state'i döndürmelidir. State ve işlem herhangi bir tür olabilir.

Reviewer: ```suggestion
* `reducer`: State'in nasıl güncelleneceğini belirleyen reducer fonksiyonudur. Saf (pure) olmalı, state'i ve aksiyonu argüman olarak almalı ve bir sonraki state'i döndürmelidir. State ve aksiyon herhangi bir tür olabilir.
```

Author: action'in aksiyon olarak cevrilmesini tavsiye eden bir kilavuz var mi? Tum action kelimelerini islem olarak cevirdim.

Reviewer 2: Yani bence ikisi de olabilir. `action.type` -> `islem tipi`. Garip gelmedi gozume. Kalabilir `islem` bence. Sadece reducerdaki `action` cok yaygin olarak issuelarda ve sorularda gectigi icin, bence parantez icinde `(action)` olarak belirtebiliriz. Bunu her yerde yapmamiza gerek yok. Action muhabbetinin ilk gectigi ve aciklandigi yer neresiylse orada belirtsek yeterli.

* `initialArg`: Başlangıç state'inin hesaplandığı değerdir. Herhangi bir türden bir değer olabilir. Başlangıç state'inin nasıl hesaplandığı, sonraki `init` argümanına bağlıdır.
* **isteğe bağlı** `init`: Başlangıç state'ini döndürmesi gereken başlatıcı fonksiyondur. Belirtilmezse, başlangıç state'i `initialArg` olarak ayarlanır. Aksi takdirde, başlangıç state'i `init(initialArg)` çağrısının sonucuna ayarlanır.

> #### Return'ler {/*returns*/}

Reviewer: ```suggestion
#### Dönüş değerleri {/*returns*/}
```

`useReducer`, tam olarak iki değer içeren bir dizi döndürür:

> 1. Mevcut state. İlk render sırasında, `init(initialArg)` veya `init` yoksa `initialArg` olarak ayarlanır.

Reviewer: ```suggestion
1. Mevcut state. İlk render sırasında, `init(initialArg)` veya `initialArg` (`init` olmadığında) olarak ayarlanır.
```

2. State'i farklı bir değere güncellemenizi ve yeniden render tetiklemenizi sağlayan [`dispatch`](#dispatch) fonksiyonu.

> #### Dikkat Edilmesi Gerekenler {/*caveats*/}

Reviewer: ```suggestion
#### Dikkat edilmesi gerekenler {/*caveats*/}
```

* `useReducer`, bir Hook olduğundan, yalnızca bileşeninizin **üst düzeyinde** veya kendi Hook'larınızda çağırabilirsiniz. Döngüler veya koşullar içinde çağıramazsınız. Buna ihtiyacınız varsa, yeni bir bileşen oluşturun ve state'i taşıyın.

> * Strict Mode'da, React, [tesadüfi karışıklıkları bulmanıza yardımcı olmak için](#my-reducer-or-initializer-function-runs-twice) reducer ve başlatıcı fonksiyonunuzu **iki kez çağırır**. Bu, yalnızca geliştirme amaçlı bir davranıştır ve canlı ortama etki etmez. Reducer ve başlatıcı fonksiyonlarınız saf halde ise (olmaları gerektiği gibi), bu mantığınızı etkilememelidir. Çağrılardan birinin sonucu yoksayılır.

Reviewer: ```suggestion
* Strict Mode'da, React, [tesadüfi karışıklıkları bulmanıza yardımcı olmak için](#my-reducer-or-initializer-function-runs-twice) reducer ve başlatıcı fonksiyonunuzu **iki kez çağırır**. Bu, yalnızca geliştirme amaçlı bir davranıştır ve canlı ortamı etkilemez. Reducer ve başlatıcı fonksiyonlarınız saf halde ise (olmaları gerektiği gibi), bu mantığınızı etkilememelidir. Çağrılardan birinin sonucu yoksayılır.
```

---

> ### `dispatch` Fonksiyonu {/*dispatch*/}

Reviewer: ```suggestion
### `dispatch` fonksiyonu {/*dispatch*/}
```

  // ...
```

React, `dispatch` fonksiyonuna ilettiğiniz eylemi ve geçerli `state` ile çağırdığınız `reducer` işlevinin sonucunu kullanarak, bir sonraki state'i ayarlayacaktır.

#### Parametreler {/*dispatch-parameters*/}

* `action`: Kullanıcı tarafından gerçekleştirilen eylem. Herhangi bir türde bir değer olabilir. Genellikle bir eylem, kendisini tanımlayan bir `type` özelliği ve isteğe bağlı olarak ek bilgi içeren diğer özellikler olan bir nesne olarak temsil edilir.

> #### Returns {/*dispatch-returns*/}

Reviewer: ```suggestion
#### Dönüş değerleri {/*dispatch-returns*/}
```

`dispatch` fonksiyonları bir dönüş değeri içermez.

> #### Dikkat Edilmesi Gereken Noktalar {/*setstate-caveats*/}

Reviewer: ```suggestion
#### Dikkat edilmesi gereken noktalar {/*setstate-caveats*/}
```

* `dispatch` fonksiyonu, sadece **bir sonraki** render işlemi için state değişkenini günceller. Eğer `dispatch` fonksiyonunu çağırdıktan sonra state değişkenini okursanız, [çağrı öncesinde ekranda olan eski değeri](#ive-dispatched-an-action-but-logging-gives-me-the-old-state-value) elde edersiniz.

> * Eğer sağladığınız yeni değer, bir [`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) karşılaştırması ile belirlendiği gibi, mevcut `state` ile aynı ise, React elemanı ve alt elamanlarının yeniden render edilmesini **atlar.** Bu bir optimizasyondur. React, sonucu yok saymadan önce bileşeninizi çağırmaya hâlâ ihtiyaç duyabilir, ancak kodunuzu etkilememelidir.

Reviewer: ```suggestion
* Eğer sağladığınız yeni değer, bir [`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) karşılaştırması ile belirlendiği gibi, mevcut `state` ile aynı ise React elemanı ve alt elemanlarının yeniden render edilmesini **atlar.** Bu bir optimizasyonudur. React, sonucu yok saymadan önce yine de bileşeninizi çağırması gerekebilir ancak bu kodunuzu etkilememelidir.
```

> * React, state güncellemelerini **toplu halde işler**. Tüm olay yöneticileri çalıştırıldıktan ve kendi `set` fonksiyonlarını çağırdıktan **sonra ekranı günceller.** Bu, tek bir olay sırasında birden fazla yeniden render işlemini önler. DOM'a erişmek için örneğin ekranı daha önce güncellemek zorunda kalmanız gereken nadir durumlarda kullanabilirsiniz. [`flushSync`](/reference/react-dom/flushSync)

Reviewer: ```suggestion
* React, state güncellemelerini **toplu halde işler**. Tüm olay yöneticileri çalıştırıldıktan ve kendi `set` fonksiyonlarını çağırdıktan **sonra ekranı günceller.** Bu, tek bir olay sırasında birden fazla yeniden render işlemini önler. React'ı ekranı güncellemeye zorlamanız gereken nadir durumlarda, örneğin DOM'a erişmek için, [`flushSync`](/reference/react-dom/flushSync) kullanabilirsiniz.
```

---

## Kullanım {/*usage*/}

> ### Bir Bileşene Reducer Eklemek {/*adding-a-reducer-to-a-component*/}

Reviewer: ```suggestion
### Bir bileşene reducer eklemek {/*adding-a-reducer-to-a-component*/}
```

> Bileşenizin state'ini [reducer](/learn/extracting-state-logic-into-a-reducer) ile yönetmek için, `useReducer` fonksiyonunu bileşeninizin en üst düzeyinde çağırın.

Reviewer: ```suggestion
Bileşeninizin state'ini [reducer](/learn/extracting-state-logic-into-a-reducer) ile yönetmek için, `useReducer`'ı bileşeninizin en üst düzeyinde çağırın.
```

</Sandpack>

`useReducer`, [`useState`](/reference/react/useState) ile çok benzerdir, ancak state güncelleme mantığını olay yöneticilerinden bileşeninizin dışındaki tek bir bir fonksiyona taşımanıza olanak tanır. [`useState` ve `useReducer` arasında seçim yapma hakkında daha fazla bilgi edinin.](/learn/extracting-state-logic-into-a-reducer#comparing-usestate-and-usereducer)

---

> ### Reducer Fonksiyonunu Yazmak {/*writing-the-reducer-function*/}

Reviewer: ```suggestion
### Reducer fonksiyonu yazmak {/*writing-the-reducer-function*/}
```

<Solution />

> #### Yapılacaklar Listesi (dizi) {/*todo-list-array*/}

Reviewer: ```suggestion
#### Yapılacaklar listesi (dizi) {/*todo-list-array*/}
```

<Solution />

> #### Immer ile Özlü Güncelleme Mantığı Yazmak {/*writing-concise-update-logic-with-immer*/}

Reviewer: ```suggestion
#### Immer ile sade güncelleme mantığı yazmak {/*writing-concise-update-logic-with-immer*/}
```

---

> ### Başlangıç State'ini Yeniden Oluşturmayı Önleme {/*avoiding-recreating-the-initial-state*/}

Reviewer: ```suggestion
### Başlangıç state'ini yeniden oluşturmayı önleme {/*avoiding-recreating-the-initial-state*/}
```

  // ...
```

Yukarıdaki örnekte, `createInitialState` bir `username` argümanı alır. Başlatıcı fonksiyonunuz başlangıç state'ini hesaplamak için herhangi bir bilgiye ihtiyacı yoksa, `null`'i `useReducer`'ın ikinci argümanı olarak geçebilirsiniz.

Dikkat edin ki, `createInitialState` **kendisi** olan *fonksiyonu* geçiriyorsunuz ve çağrı sonucu olan `createInitialState()`'i değil. Bu şekilde, başlatma işleminden sonra başlangıç state'i yeniden oluşturulmaz.

<Recipes titleText="Başlatıcı fonksiyonu geçirmenin ve başlangıç state'ini doğrudan geçirmenin farkı" titleId="examples-initializer">

> #### Başlatıcı Fonksiyonunu Geçirme {/*passing-the-initializer-function*/}

Reviewer: ```suggestion
#### Başlatıcı fonksiyonu geçirme {/*passing-the-initializer-function*/}
```

<Solution />

> #### Başlangıç State'ini Doğrudan Geçirme {/*passing-the-initial-state-directly*/}

Reviewer: ```suggestion
#### Başlangıç state'ini doğrudan geçirme {/*passing-the-initial-state-directly*/}
```

---

> ## Sorun Giderme {/*troubleshooting*/}

Reviewer: ```suggestion
## Sorun giderme {/*troubleshooting*/}
```

> ### Bir işlem yaptım, ancak günlüklemede bana eski state değerini veriyor {/*ive-dispatched-an-action-but-logging-gives-me-the-old-state-value*/}

Reviewer: ```suggestion
### Bir aksiyon tetikledim ancak state'i yazdırdığımda eski değerini veriyor {/*ive-dispatched-an-action-but-logging-gives-me-the-old-state-value*/}
```

---

> ### Bir işlem yaptım, ancak ekran güncellenmiyor {/*ive-dispatched-an-action-but-the-screen-doesnt-update*/}

Reviewer: ```suggestion
### Bir aksiyon tetikledim ancak ekran güncellenmiyor {/*ive-dispatched-an-action-but-the-screen-doesnt-update*/}
```

  throw Error('Bilinmeyen eylem:: ' + action.type);
}
```

Böyle hataları yakalamak için TypeScript gibi bir statik tip denetleyicisi de kullanabilirsiniz.

---

### "Too many re-renders" hatası alıyorum {/*im-getting-an-error-too-many-re-renders*/}

`Too many re-renders. React limits the number of renders to prevent an infinite loop.` hatası alabilirsiniz. Bu genellikle, işlemi koşulsuz bir şekilde `render` sırasında gönderdiğiniz anlamına gelir, böylece bileşeniniz döngüye girer: render, gönderim (bu da bir yeniden render yapar), render, gönderim (bu da bir yeniden render yapar) ve böyle devam eder. Bu sıklıkla, bir olay yöneticisi belirleme hatası nedeniyle oluşur:

```js {1-2}
// 🚩 Yanlış: yöneticiyi yeniden render sırasında çağırır.
return <button onClick={handleClick()}>Tıkla</button>

// ✅ Doğru: olay yöneticisini aşağıya aktarır.
return <button onClick={handleClick}>Tıkla</button>

// ✅ Doğru: iç içe bir fonksiyonu aktarır.
return <button onClick={(e) => handleClick(e)}>Tıkla</button>
```

Bu hatanın nedenini bulamazsanız, konsoldaki hatanın yanındaki ok'a tıklayın ve JavaScript yığınını (stack) tarayarak hatadan sorumlu belirli `dispatch` fonksiyonu çağrısını bulun.

---

> ### Reducer veya Başlatıcı (initializer) Fonksiyonlarım İki Kez Çalışıyor. {/*my-reducer-or-initializer-function-runs-twice*/}

Reviewer: ```suggestion
### Reducer veya başlatıcı (initializer) fonksiyonlarım iki kez çalışıyor {/*my-reducer-or-initializer-function-runs-twice*/}
```
